### PR TITLE
Fix Android IME composition history recovery

### LINKED ISSUE #828: Android IME corruption still drops first letters and spaces while editing
## Problem

The Android/mobile editor corruption bug is still reproducible after the previous fixes.

Strong reproduction from real-device use:
- in an empty blip, typing `n` appears first
- when typing `e`, the earlier `n` disappears
- same thing happens again at the start of the second word (`b` in `blip`)
- pressing Done removes the space between words
- final persisted result becomes `ewlip` instead of `new blip`

This means the prior mobile edit-loss and IME fallback fixes were incomplete for the real Android composition path.

## Scope

- root-cause-first reinvestigation of the remaining Android IME corruption path
- do not assume the prior fixes were sufficient
- identify the exact failing boundary for character replacement / missing first-letter / missing-space corruption
- implement the minimal correct fix
- verify locally as far as possible and document any emulator vs. real-device limits
- open PR when clean

## Required regression target

At minimum preserve the exact user-level behavior:
- entering `new blip` in an empty blip must persist exactly `new blip`
- no dropped first characters
- no collapsed spaces on Done/teardown

### PR DESCRIPTION
**Pull request by vega113**
Fixes #828.

## Summary
- Reconstruct Android IME replacement streams from `compositionupdate.data`, covering the observed `n -> e -> ew` and `b -> l -> li -> lip` sequences that dropped first letters in Chrome on Galaxy devices.
- Insert standalone Android `textInput` separators directly when they arrive outside an active composition, so spaces are not left as unmanaged DOM mutations.
- Suppress stale empty-baseline ghost restoration after the recovered composition text is flushed, avoiding the blank blip seen after tapping Done.

## Verification
- `sbt "Test/runMain org.junit.runner.JUnitCore org.waveprotocol.wave.model.util.ImeCompositionTextTrackerTest org.waveprotocol.wave.model.util.GhostTextReconcilerTest"` - PASS, OK (49 tests)
- `sbt compileGwtDev` - PASS, GWT one-permutation draft compile succeeded
- `python3 scripts/assemble-changelog.py && python3 scripts/validate-changelog.py --fragments-dir wave/config/changelog.d --changelog wave/config/changelog.json` - PASS, assembled 243 entries and validation passed
- `bash scripts/worktree-boot.sh --port 9933` - PASS, staged app/runtime built
- `PORT=9933 bash scripts/wave-smoke.sh check` - PASS, root/health/landing/J2CL/webclient endpoints returned expected 200 statuses
- Galaxy S9+ Chrome mobile simulation against local GWT server: typed `new blip`, tapped Done, reloaded, and verified `new blip` remained visible after each step
- `git diff --check` - PASS

## Summary by CodeRabbit

* **Bug Fixes**
  * Android Chrome IME composition now reconstructs composed words, preserves leading letters/spaces, and commits standalone Android text separators so typed phrases (e.g., “new blip”) persist after Done and reload.

* **Tests**
  * Adds unit and integration tests covering composition recovery, suppression edge cases, and end-to-end persistence verification.

* **Documentation**
  * Adds a documentation plan and changelog entry describing the behavior and required verification steps.

### CODE REVIEW

File: wave/src/main/java/org/waveprotocol/wave/model/util/ImeCompositionTextTracker.java [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+  private String lastObserved = "";
+
+  public void reset() {
+    reconstructed = "";
+    lastObserved = "";
+  }
+
+  public void observe(String value) {
+    if (value == null || value.isEmpty()) {
+      return;
```
[the remark this answers is not in the document]
Comment (vega113) [reply wave/src/main/java/org/waveprotocol/wave/model/util/ImeCompositionTextTracker.java]: Fixed in 0e348c31e: observe() now calls reset() on empty input, so stale reconstructed text is cleared and effectiveText returns empty. Added testEmptyUpdateClearsRecoveredText.

File: wave/src/main/java/org/waveprotocol/wave/model/util/ImeCompositionTextTracker.java [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+      return scratch;
+    }
+    if (scratch.isEmpty()) {
+      return reconstructed;
+    }
+    if (scratch.endsWith(reconstructed)) {
+      return scratch;
+    }
+    if (reconstructed.endsWith(scratch)) {
+      return reconstructed;
```
[the remark this answers is not in the document]
Comment (vega113) [reply wave/src/main/java/org/waveprotocol/wave/model/util/ImeCompositionTextTracker.java]: Fixed in 0e348c31e: effectiveText() now gates the reconstructed-wins branch on hasRecoveredReplacement and scratch.equals(lastObserved), so a shrink like abc->bc correctly keeps the live scratch. Added testShrinkWithoutReplacementEvidenceKeepsScratch.

File: wave/src/main/java/org/waveprotocol/wave/model/util/ImeCompositionTextTracker.java [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+
+  public String effectiveText(String scratchText) {
+    String scratch = scratchText == null ? "" : scratchText;
+    if (reconstructed.isEmpty()) {
+      return scratch;
+    }
+    if (scratch.isEmpty()) {
+      return reconstructed;
+    }
+    if (scratch.endsWith(reconstructed)) {
```
[the remark this answers is not in the document]
Comment (vega113) [reply wave/src/main/java/org/waveprotocol/wave/model/util/ImeCompositionTextTracker.java:110]: Fixed in 0e348c31e: empty observe() now calls reset(), and effectiveText() requires hasRecoveredReplacement before preferring reconstructed. Both cases have new unit test coverage.

**Comment by vega113** [comment]
Applied all three CodeRabbit nitpick suggestions in commit 3548c644a:

1. **SignalEvent.getData() Javadoc** — Updated to specify non-null return semantics: "never null — implementations return `""` when there is no text payload.".
2. **CompositionEventHandlerTest.verifyUpdate()** — Refactored to accept varargs token parameter (matching the existing `verifyStart` pattern), so future tests can assert specific tokens without changing existing call sites.
3. **ImeCompositionTextTrackerTest** — Extracted `assertImeSequence()` helper; refactored all seven test methods to use it, reducing ~52 lines of boilerplate.

File: wave/src/main/java/org/waveprotocol/wave/model/util/ImeCompositionTextTracker.java [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+    }
+    if (!lastObserved.isEmpty() && value.startsWith(lastObserved)) {
+      reconstructed += value.substring(lastObserved.length());
+      lastObserved = value;
+      return;
+    }
+    if (isSingleCharacterReplacement(value)) {
+      reconstructed += value;
+      lastObserved = value;
+      hasRecoveredReplacement = true;
```
[the remark this answers is not in the document]
Comment (vega113) [reply wave/src/main/java/org/waveprotocol/wave/model/util/ImeCompositionTextTracker.java]: Fixed in 3556a3236: single-character replacements are now held as pending until a duplicate or extension confirms the Android replacement stream, and replacement recovery is limited to ASCII key streams. Diacritic revisions such as a -> \u00e1 now fall back to the live scratch. Added testSingleCharacterDiacriticReplacementKeepsScratch plus a pending-extension recovery test.

[the remark this answers is not in the document]
Comment (vega113) [reply wave/src/main/java/org/waveprotocol/wave/model/util/ImeCompositionTextTracker.java]: Fixed in 3556a3236: `isSingleCharacterReplacement` now gates on `isAsciiLetterOrDigit`, excluding diacritic characters (e.g. `á`) from the word-building path entirely. A pending-replacement pattern delays committing the appended char until a duplicate or extension confirms it is Android Chrome word-building behaviour.

File: wave/src/test/java/org/waveprotocol/wave/model/util/ImeCompositionTextTrackerTest.java [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+ */
+public class ImeCompositionTextTrackerTest extends TestCase {
+
+  private void assertImeSequence(String expectedText, String scratchText, String... observedValues) {
+    ImeCompositionTextTracker tracker = new ImeCompositionTextTracker();
+    for (String v : observedValues) {
+      tracker.observe(v);
+    }
+    assertEquals(expectedText, tracker.effectiveText(scratchText));
+  }
```
[the remark this answers is not in the document]
Comment (vega113) [reply wave/src/test/java/org/waveprotocol/wave/model/util/ImeCompositionTextTrackerTest.java:36]: Tests added in the current branch HEAD: `testNullObservedValueIsIgnored` (observe null after observe("n") keeps state) and `testNullScratchFallsBackToEmptyString` (effectiveText(null) returns empty string when scratch is null). Both null-guard paths are now covered.

File: wave/src/main/java/org/waveprotocol/wave/client/editor/event/EditorEventHandler.java [hunk 6]
```diff
@@ -534,11 +537,52 @@ private void compositionStart(EditorEvent event) {
...
+          .emit();
+    }
+    editorInteractor.compositionUpdate(event.getData());
+  }
+
+  private boolean shouldHandleStandaloneAndroidTextInput(EditorEvent event) {
+    return state == State.NORMAL
+        && UserAgent.isAndroid()
+        && BrowserEvents.TEXTINPUT.equals(event.getType())
+        && event.getData() != null
```
[the remark this answers is not in the document]
Comment (vega113) [reply wave/src/main/java/org/waveprotocol/wave/client/editor/event/EditorEventHandler.java:561]: Fixed in e3ada6549: the standalone Android textInput path now requires a just-ended Android IME composition before it can insert data directly. That expectation survives IME keyCode 229 follow-up events, is cleared by unrelated events, empty/non-handled textInput, compositionstart, or after one handled textInput. This keeps the user-log space after compositionend covered while avoiding a broad Android textInput double-insert path. Local compile checks and focused tracker/ghost tests passed before push.

[the remark this answers is not in the document]
Comment (vega113) [reply wave/src/main/java/org/waveprotocol/wave/client/editor/event/EditorEventHandler.java:561]: Fixed in e3ada65: added `androidImeStandaloneTextInputExpected` flag that is set only when a preceding IME keydown (keyCode 229) is observed and cleared on any non-IME or non-textInput event. This gates the standalone TEXTINPUT path exclusively to IME-originated input, preventing double-insertion from hardware keyboard events.

File: wave/src/main/java/org/waveprotocol/wave/model/util/ImeCompositionTextTracker.java [hunk 7]
```diff
@@ -0,0 +1,143 @@
...
+      } else {
+        reconstructed += value.substring(lastObserved.length());
+      }
+      lastObserved = value;
+      return;
+    }
+    if (!pendingReplacementValue.isEmpty()
+        && value.length() == 1
+        && isAsciiLetterOrDigit(value.charAt(0))) {
+      confirmPendingReplacement();
```
[the remark this answers is not in the document]
Comment (vega113) [reply wave/src/main/java/org/waveprotocol/wave/model/util/ImeCompositionTextTracker.java]: Fixed in 3ede9aef0: when a new single-char ASCII arrives while a pending replacement exists, the pending value is now updated in-place and the original base is kept — intermediate unconfirmed chars are never accumulated. The stream "a → b → c → c" now yields "ac" instead of "abc". Added testConsecutiveSingleCharReplacementsAvoidStalePrefix to prevent regression.
